microcosm_elasticsearch/registry.py

```diff
@@ -0,0 +1,80 @@
+"""
+Manage a set of indexes and/or aliases.
+
+"""
+from elasticsearch_dsl import Index
+from inflection import underscore
+
+
+class IndexRegistry(object):
+    """
+    A registry of application indexes.
+
+    """
+    def __init__(self, graph):
+        self.graph = graph
+        self.indexes = {}
+
+    def register(self, name=None, version=None):
+        """
+        Register an index locally.
+
+        Note that `createall` is needed to save the index to Elasticsearch.
+
+        The index will be named per convention such that:
+         -  The graph's name is used by default
+         -  The "test" suffix is added for unit testing (to avoid clobbering real data)
+
+        If version is provided, it will be used to create generate an alias (to the unversioned name).
+
+        """
+        if version is None:
+            index_name = IndexRegistry.name_for(self.graph, name=name)
+            alias_name = None
+        else:
+            # create index with full version, alias to shortened version
+            index_name = IndexRegistry.name_for(self.graph, name=name, version=version)
+            alias_name = IndexRegistry.name_for(self.graph, name=name)
+
+        if index_name in self.indexes:
+            raise Exception("Index already registered for name: {}".format(index_name))
+
+        index = Index(
+            name=index_name,
+            using=self.graph.elasticsearch_client,
+        )
+        if alias_name is not None:
```

Comment:
ES supports aliases to indexes. In this case, if we pass a version (e.g. `v1`), we create the index named `foo-v1` and alias it to `foo`. This lets us evolve the index mappings without changing references to the index in query code.

I have in mind something Alembic-like that uses a hash/uuid as the version and automatically runs the reindex API...